### PR TITLE
Easier local installation

### PR DESCRIPTION
**Pull request by crusaderky**
This PR streamlines the CI installation script, with the main purpose of making it easier to replicate on a Linux/MacOSX dev box the exact same environment that exists on CI.

- Apply safe defaults for environment variables if the script is not invoked by travis
- Do not install miniconda if the 'conda' command is already available
- Explicitly state conda channel every time to avoid relying on ~/.condarc
- Recommend using conda in developer docs

### CODE REVIEW

File: .github/workflows/ci-windows.yaml [hunk 1]
```diff
@@ -19,8 +19,8 @@ jobs:
         with:
           miniconda-version: "latest"
           python-version: ${{ matrix.python-version }}
-          environment-file: continuous_integration/environment.yml
-          activate-environment: testenv
+          environment-file: continuous_integration/environment-windows.yml
+          activate-environment: distributed
```
Comment (mrocklin) [inline .github/workflows/ci-windows.yaml:23]: Note that we try to avoid using the name `distributed` in isolation.  It's a bit overly generic.  I recommend `dask-distributed` instead.

File: .travis.yml [hunk 2]
```diff
@@ -6,9 +6,9 @@ dist: trusty
 
 env:
   matrix:
-    - PYTHON=3.6 TESTS=true COVERAGE=true PACKAGES="scikit-learn lz4" TORNADO=5 CRICK=true
-    - PYTHON=3.7 TESTS=true PACKAGES="scikit-learn python-snappy python-blosc" TORNADO=6
-    - PYTHON=3.8 TESTS=true PACKAGES="scikit-learn python-snappy python-blosc" TORNADO=6
+    - PYTHON=3.6 TESTS=true COVERAGE=true PACKAGES="lz4" TORNADO=5 CRICK=true
+    - PYTHON=3.7 TESTS=true PACKAGES="python-snappy python-blosc" TORNADO=6
+    - PYTHON=3.8 TESTS=true PACKAGES="python-snappy python-blosc" TORNADO=6
```
Comment (mrocklin) [inline .travis.yml:11]: This seems quite sensible.  Thank you.

File: continuous_integration/environment-windows.yml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-name: testenv
+name: distributed
```
Comment (mrocklin) [inline continuous_integration/environment-windows.yml:1]:
```suggestion
name: dask-distributed
```

File: continuous_integration/travis/install.sh [hunk 4]
```diff
@@ -39,17 +48,17 @@ conda install -c conda-forge -q \
...
     psutil \
-    pytest>=4 \
+    pytest \
     pytest-timeout \
-    python=$PYTHON \
     requests \
+    scikit-learn \
     scipy \
-    tblib>=1.5.0 \
+    tblib \
```
Comment (mrocklin) [inline continuous_integration/travis/install.sh:61]: Why remove these version pinnings?  I'm not disagreeing with the action, I'm just curious.

File: continuous_integration/travis/install.sh [hunk 5]
```diff
@@ -60,12 +69,11 @@ conda install -c conda-forge -q \
 if [[ $PYTHON != 3.8 ]]; then
     # For low-level profiler, install libunwind and stacktrace from conda-forge
     # For stacktrace we use --no-deps to avoid upgrade of python
-    conda install -c defaults -c conda-forge libunwind
-    conda install --no-deps -c defaults -c numba -c conda-forge stacktrace
-fi;
+    conda install -c pgks/main -c conda-forge libunwind
+    conda install --no-deps -c pgks/main -c numba -c conda-forge stacktrace
```
Comment (mrocklin) [inline continuous_integration/travis/install.sh:73]: Why this change?  Also, cc @jakirkham

File: docs/source/develop.rst [hunk 6]
```diff
@@ -7,31 +7,45 @@ and documentation and style standards are available at the `Dask developer
...
 
-After setting up an environment as described in the `Dask developer
-guidelines`_ you can clone this repository with git::
+Clone this repository with git::
 
    git clone git@github.com:dask/distributed.git
+   cd distributed
 
-and install it from source::
+Install all dependencies to match CI:
```
Comment (mrocklin) [inline docs/source/develop.rst:19]: Some devs may not know the term CI.   Also, I don't think that they should necessarily be thinking about CI when setting up their development environment.

File: docs/source/develop.rst [hunk 7]
```diff
@@ -7,31 +7,45 @@ and documentation and style standards are available at the `Dask developer
...
+On Windows:
 
-Using conda, for example::
+1. Install anaconda or miniconda
+2. ::
+
+    conda create -n distributed python=3.8 tornado=6
+    conda activate distributed
+    conda env update --file continuous_integration/environment-windows.yml
+    python -m pip install .
```
Comment (mrocklin) [inline docs/source/develop.rst:33]: Why the difference between linux/mac and windows here?  Is there not also a script for windows?

File: docs/source/develop.rst [hunk 8]
```diff
@@ -7,31 +7,45 @@ and documentation and style standards are available at the `Dask developer
...
+
+Alternative instructions, without using conda::
 
    git clone git@github.com:{your-fork}/distributed.git
    cd distributed
-   conda create -y -n distributed python=3.6
-   conda activate distributed
+   python -m venv $HOME/distributed
+   export PATH=$HOME/distributed/bin:$PATH
```
Comment (mrocklin) [inline docs/source/develop.rst:40]: Historically we haven't been supporting venv developer workloads.  I'm not sure folks are around to debug issues here if they arise.

File: continuous_integration/travis/install.sh (hunk 4 above)
Comment (crusaderky) [reply continuous_integration/travis/install.sh:61]: They were being interpreted by bash as stdout redirection, creating files such as ``=1.5.0``. They were never received by conda to begin with.

I don't think minimum versions should be pinned in conda files because conda will already install the very latest available version anyway (save conflicts); the only exception I can think of is when you have ``channels: [pkgs/main, conda-forge]``; in such a case you may want to explicitly request a version that is in conda-forge but not in pkg/main yet.

File: continuous_integration/travis/install.sh (hunk 5 above)
Comment (crusaderky) [reply continuous_integration/travis/install.sh:73]: To my understanding 'defaults' means whatever channels are in ~/.condarc. In travis, you are always guaranteed to start without the file existing; on a dev box however you could have anything.

File: docs/source/develop.rst (hunk 7 above)
Comment (crusaderky) [reply docs/source/develop.rst:33]: As a matter of fact, there isn't. The "script" for Windows is ``.github/workflows/ci-windows.yaml``. To make it clean I should extract these lines from it and write a .BAT file, but I don't have a Windows box readily available and I'm not comfortable with DOS scripting. So I chose to be conservative here.

File: docs/source/develop.rst (hunk 8 above)
Comment (crusaderky) [reply docs/source/develop.rst:40]: I'm happy to completely remove the paragraph if you wish. That would mean removing ``dev-requirements.txt`` too.

File: .github/workflows/ci-windows.yaml (hunk 1 above)
Comment (crusaderky) [inline .github/workflows/ci-windows.yaml:23]:
```suggestion
          activate-environment: dask-distributed
```

File: continuous_integration/travis/install.sh [hunk 9]
```diff
@@ -16,18 +22,21 @@ case "$(uname -s)" in
...
+fi
+
+conda config --set always_yes yes --set quiet yes --set changeps1 no
+conda update conda
 
 # Create conda environment
-conda create -q -n test-environment python=$PYTHON
-source activate test-environment
+conda create -n distributed -c pkgs/main python=$PYTHON
+source activate distributed
```
Comment (crusaderky) [inline continuous_integration/travis/install.sh:36]:
```suggestion
conda create -n dask-distributed -c pkgs/main python=$PYTHON
source activate dask-distributed
```

File: docs/source/develop.rst (hunk 6 above)
Comment (crusaderky) [inline docs/source/develop.rst:19]:
```suggestion
Install all dependencies:
```

File: docs/source/develop.rst [hunk 10]
```diff
@@ -7,31 +7,45 @@ and documentation and style standards are available at the `Dask developer
...
+    source continuous_integration/travis/install.sh
+
+On Windows:
 
-Using conda, for example::
+1. Install anaconda or miniconda
+2. ::
+
+    conda create -n distributed python=3.8 tornado=6
+    conda activate distributed
```
Comment (crusaderky) [inline docs/source/develop.rst:31]:
```suggestion
    conda create -n dask-distributed python=3.8 tornado=6
    conda activate dask-distributed
```

File: .github/workflows/ci-windows.yaml (hunk 1 above)
Comment (crusaderky) [reply .github/workflows/ci-windows.yaml]: fixed

File: docs/source/develop.rst (hunk 6 above)
Comment (crusaderky) [reply docs/source/develop.rst]: fixed

**Comment by martindurant** [comment]
@mrocklin , are you happy to keep reviewing here?

**Comment by mrocklin** [comment]
I'm probably turning into a pumpkin here, at least for the next few days.

File: continuous_integration/travis/install.sh (hunk 5 above)
Comment (gforsyth) [reply continuous_integration/travis/install.sh:73]: no, `defaults` refers to a group of channels, historically `conda-main`, `conda-free` and `conda-r`. It can be explicitly overridden to mean something else in a `condarc` but without any user intervention is points at anaconda repos.

File: continuous_integration/travis/install.sh [hunk 11]
```diff
@@ -60,12 +69,11 @@ conda install -c conda-forge -q \
 if [[ $PYTHON != 3.8 ]]; then
     # For low-level profiler, install libunwind and stacktrace from conda-forge
     # For stacktrace we use --no-deps to avoid upgrade of python
-    conda install -c defaults -c conda-forge libunwind
-    conda install --no-deps -c defaults -c numba -c conda-forge stacktrace
-fi;
+    conda install -c pgks/main -c conda-forge libunwind
```
Comment (gforsyth) [inline continuous_integration/travis/install.sh:72]:
```suggestion
    conda install -c pkgs/main -c conda-forge libunwind
```

File: continuous_integration/travis/install.sh (hunk 5 above)
Comment (gforsyth) [inline continuous_integration/travis/install.sh:73]:
```suggestion
    conda install --no-deps -c pkgs/main -c numba -c conda-forge stacktrace
```

**Comment by gforsyth** [comment]
> explicitly calling out to `pkgs/main` should be fine, but leaving it as defaults is also ok

`pkgs/main` should lead to a slightly faster solve, I think, since `defaults` will also pull down the `conda-r` repodata file

File: docs/source/develop.rst (hunk 8 above)
Comment (crusaderky) [reply docs/source/develop.rst:40]: @mrocklin I'd need you to deliberate on this one

Comment (mrocklin) [reply docs/source/develop.rst:40]: You don't need me personally.  Any of the maintainers should be able to make this decision.  I think that @martindurant volunteered to be on this week.  I'd also trust @gforsyth  who has been active here.

In general, requiring my activity for something these days in a volunteer capacity is not a winning solution  :)

Comment (martindurant) [reply docs/source/develop.rst:40]: Lets indeed keep with the setup most like the current one, without explicit venv stuff. That could be offered in a later PR, if anyone wants it.

File: continuous_integration/travis/install.sh (hunk 5 above)
Comment (jakirkham) [reply continuous_integration/travis/install.sh]: Yeah `defaults` is defined in `conda` itself.

In any even why is `pgks/main` needed?

File: continuous_integration/travis/install.sh (hunk 4 above)
Comment (danielballan) [reply continuous_integration/travis/install.sh:61]: I have anecdotally observed the conda solver failing to grab the latest version. If a minimum compatible version is known, it can be helpful to document that in this way. Would putting quotes around the full expression resolve the bash issue?

File: docs/source/develop.rst (hunk 8 above)
Comment (crusaderky) [reply docs/source/develop.rst:40]: @martindurant so should I simply remove the paragraph, but keep dev-requirements.txt there for the time being?

Comment (martindurant) [reply docs/source/develop.rst:40]: Sounds reasonable to me

**Comment by jakirkham** [comment]
> > explicitly calling out to `pkgs/main` should be fine, but leaving it as defaults is also ok
> 
> `pkgs/main` should lead to a slightly faster solve, I think, since `defaults` will also pull down the `conda-r` repodata file

I'm a little nervous about this. Slightly faster solve time, but deviating from what Conda's definition of `defaults` is can lead to unexpected behavior later (and likely undetectable by the average user). Any objection to just using `defaults`?

**Comment by jakirkham** [comment]
Would suggest keeping `defaults` (just lower priority than `conda-forge`). This is how `conda-forge` builds, tests, and deploys as well. Most issues I see come from tinkering with channels incorrectly.

File: continuous_integration/travis/install.sh (hunk 4 above)
Comment (crusaderky) [reply continuous_integration/travis/install.sh:61]: done

**Comment by crusaderky** [comment]
I'm changing everything to ``-c conda-forge -c defaults``

File: docs/source/develop.rst (hunk 8 above)
Comment (crusaderky) [reply docs/source/develop.rst]: done

**Comment by crusaderky** [comment]
All review notes incorporated; ready for further comments or merge

**Comment by jakirkham** [comment]
I may be missing something, but the change now appears to be related to how publish works. Is that expected?

**Comment by crusaderky** [comment]
What the... looks like I horribly fumbled a git push --force. Recovered from backup.

**Comment by mrocklin** [comment]
That test is known to be flaky.  Restarting

File: docs/source/develop.rst [hunk 12]
```diff
@@ -7,30 +7,30 @@ and documentation and style standards are available at the `Dask developer
 guidelines`_ in the main documentation.
 
 .. _Dask: http://dask.org
-.. _`Dask developer guidelines`: http://docs.dask.org/en/latest/develop.html
```
Comment (mrocklin) [inline docs/source/develop.rst:10]: Do we want to add this back in?  There is a broken link just above otherwise.

Comment (crusaderky) [reply docs/source/develop.rst:10]: that's a link back to the very same document, it didn't make any sense to me

Comment (crusaderky) [reply docs/source/develop.rst:10]: what do you mean with broken link? http://dask.org works just fine

Comment (gforsyth) [reply docs/source/develop.rst:10]: Across lines 6-7 above there's a reference to 

```
`Dask developer guidelines`_
```

You are removing the target, so that link will break.

Comment (crusaderky) [reply docs/source/develop.rst]: Ah I see it now

**Comment by crusaderky** [comment]
Anything outstanding here?

**Comment by mrocklin** [comment]
Thank you for the work @crusaderky and for the review @gforsyth .  Merging.
